pkg/shardddl/optimism/lock.go

```diff
@@ -141,94 +114,154 @@ func (l *Lock) TrySync(callerSource, callerSchema, callerTable string,
-	//   4. DM-worker-A replicate `ADD COLUMN` to the downstream.
-	// in order to support DML from DM-worker-B matches the downstream schema,
-	// two strategies exist:
-	//   A. DM-worker-B waits for DM-worker-A to finish the replication of the DDL before replicating DML.
-	//   B. DM-worker-B also replicates the DDL before replicating DML,
-	//      but this MUST ensure we can tolerate replicating the DDL multiple times.
-	// for `DROP COLUMN` or other DDL which makes the schema become smaller,
-	// this is not a problem because all DML with larger schema should already replicated to the downstream,
-	// and any DML with smaller schema can fit both the larger or smaller schema.
-	// To make it easy to implement, we will temporarily choose strategy-B.
-
-	cmp, _ = oldTable.Compare(newTable) // we have checked `err` returned above.
-	if cmp < 0 {
-		// let every table to replicate the DDL.
-		return ddls, nil
-	} else if cmp > 0 {
-		// do nothing except the last table.
-		return emptyDDLs, nil
+	// handle the case where <callerSource, callerSchema, callerTable>
+	// is not in old source tables and current new source tables.
+	// duplicate append is not a problem.
+	tts = append(tts, newTargetTable(l.Task, callerSource, l.DownSchema, l.DownTable,
+		map[string]map[string]struct{}{callerSchema: {callerTable: struct{}{}}}))
+	// add any new source tables.
+	l.addTables(tts)
+	if val, ok := l.versions[callerSource][callerSchema][callerTable]; !ok || val < infoVersion {
+		l.versions[callerSource][callerSchema][callerTable] = infoVersion
 	}
 
-	// compare the current table's info with joined info.
-	cmp, err = newTable.Compare(newJoined)
+	var emptyDDLs = []string{}
+	oldTable := l.tables[callerSource][callerSchema][callerTable]
+	oldJoined := l.joined
+
+	// update table info and joined info base on the last new table info
+	lastTableInfo := schemacmp.Encode(newTIs[len(newTIs)-1])
+	log.L().Info("update table info", zap.String("lock", l.ID), zap.String("source", callerSource), zap.String("schema", callerSchema), zap.String("table", callerTable),
+		zap.Stringer("from", oldTable), zap.Stringer("to", lastTableInfo), zap.Strings("ddls", ddls))
+	l.tables[callerSource][callerSchema][callerTable] = lastTableInfo
+
+	lastJoined, err := joinTable(lastTableInfo)
 	if err != nil {
-		return emptyDDLs, terror.ErrShardDDLOptimismTrySyncFail.Delegate(
-			err, l.ID, "can't compare table info (new table info) %s with (new joined table info) %s", newTable, newJoined) // NOTE: this should not happen.
+		return emptyDDLs, err
 	}
-	if cmp < 0 {
-		// no need to replicate DDLs, because has a larger joined schema (in the downstream).
-		// FIXME: if the previous tables reached the joined schema has not replicated to the downstream,
-		// now, they should re-try until replicated successfully, try to implement better strategy later.
-		return emptyDDLs, nil
+	// update the current joined table info, if it's actually changed it should be logged in `if cmp != 0` block.
```

Comment:
(this PR doesn't change this line) I can't understand “if it's actually changed
”, could someone make it more clear😂